File: algo_2021/w5.-traverse-divide-and-conquer/16.Permutations_II/16_permutation__iteration.py
class Solution:
    """
    @param: :  A list of integers
    @return: A list of unique permutations
    """

    def permuteUnique(self, nums):
        # write your code here
        # write your code here
        res = []
        if nums is None:
            return res

        if len(nums) == 0:
            res.append(list(res))
            return res
        
        nums.sort()
        self.dfs(nums, set(), [], res)
        return res
    
    def dfs(self, nums, visited, permutation, res):
        if len(permutation) == len(nums):
            res.append(list(permutation))
            return 
        
        for i in range(len(nums)):
            if i in visited:
                continue
            if i > 0 and nums[i] == nums[i - 1] and i - 1 not in visited:
                continue
            visited.add(i)
            permutation.append(nums[i])
            self.dfs(nums, visited, permutation, res)
            permutation.pop()
            visited.remove(i)

File: algo_2021/w5.-traverse-divide-and-conquer/16.Permutations_II/16_permutation__iteration.java
public class Solution {
    /*
     * @param :  A list of integers
     * @return: A list of unique permutations
     */
    public List<List<Integer>> permuteUnique(int[] nums) {
        // write your code here
    List<List<Integer>> res = new ArrayList<>();
        if (nums == null) {
            return res;
        }

        if (nums.length == 0) {
            res.add(new ArrayList<Integer>());
            return res;
        }
         
        Arrays.sort(nums);
        boolean[] visited = new boolean[nums.length];
        dfs(nums, visited, new ArrayList<Integer>(), res);
        return res;
    }

    private void dfs(int[] nums,
                     boolean[] visited, 
                     List<Integer> permutation,
                     List<List<Integer>> res
                    ) {
        if (permutation.size() == nums.length) {
            res.add(new ArrayList<Integer>(permutation));
            return;
        }

        for (int i = 0; i < nums.length; i++) {
            if (visited[i]) {
                continue;
            }
            if (i > 0 && nums[i] == nums[i - 1] && visited[i - 1] == false) {
                continue;
            }
            visited[i] = true;
            permutation.add(nums[i]);
            dfs(nums, visited, permutation, res);
            permutation.remove(permutation.size() - 1);
            visited[i] = false;
        }
    }  
}